### PR TITLE
Added envoy-sync-receive workflow

### PR DESCRIPTION
**Pull request by tedjpoole**
Added `envoy-sync-receive.yaml` workflow. This is designed to be invoked from the `envoy-sync.yaml` workflow from the upstream `envoyproxy/envoy` repository (see envoyproxy/envoy#34319).

It performs a merge from the specified branch in upstream `envoyproxy/envoy` into the same named branch in `envoyproxy/envoy-openssl` e.g.
- envoyproxy/envoy[release/v1.28] -> envoyproxy/envoy-openssl[release/v1.28]
- envoyproxy/envoy[release/v1.30] -> envoyproxy/envoy-openssl[release/v1.30]
- envoyproxy/envoy[main] -> envoyproxy/envoy-openssl[main]

For each distinct branch name, the script manages at most 1 associated pull request and/or issue, depending on the success of the merge.

If the merge from upstream succeeds, it creates or updates the associated pull request, and closes the associated issue (if there is one).
If the merge from upstream fails, it leaves the associated pull request untouched (if there is one), and creates or updates the associated issue.

### CODE REVIEW

**Review by phlax — COMMENTED** [review]
looks good @tedjpoole, but will need a bit of setup with the repo/s and/or bot/s

File: ci/envoy-sync-receive.sh [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#  - pushes the feature branch to the fork
+#  - creates the associated pull request if it doesn't already exist
+#  - closes the associated issue if it already exists
+# -If the merge is unsuccessful, it:
+#  - leaves the associated pull request untouched if it already exists
+#  - creates the associated issue issue if it doesn't already exist
+#  - adds a comment on the associated issue to describe the merge fail
+#
+
+set -xeuo pipefail
```
[the remark this answers is not in the document]
Comment (tedjpoole) [reply ci/envoy-sync-receive.sh:22]: I left the `-x` on the assumption that (1) there probably will be some bugs to fix, and (2) the `-x` will make it easier to fix those bugs.

File: ci/envoy-sync-receive.sh [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+# Create a new branch name for the merge. Deliberately don't include
+# any commit hash or timestamp in the name to ensure it is repeatable.
+# This ensures that each time we get invoked, due to an upstream change,
+# we accumulate the changes in the same branch and pull request, rather
+# than creating new ones that superceed the old one(s) each time.
+DST_NEW_BRANCH_NAME="auto-merge-$(echo "${SRC_BRANCH_NAME}" | tr /. -)"
+
+# Set the default remote for the gh command
+gh repo set-default "${DST_REPO_ORG}/${DST_REPO_NAME}"
```
[the remark this answers is not in the document]
Comment (phlax) [reply ci/envoy-sync-receive.sh:76]: hmm, actually good point, if we are using the `GITHUB_TOKEN` (and setting relevant creds) we dont need the appauth stuff which essentially provides write creds

File: ci/envoy-sync-receive.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (jwendell) [reply ci/envoy-sync-receive.sh:23]: We need to pay double attention to not leak any secret/token.

File: ci/envoy-sync-receive.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (tedjpoole) [reply ci/envoy-sync-receive.sh:77]: Yes, the `gh` command gets its auth token from the `GITHUB_TOKEN` (or `GH_TOKEN`) environment variable (see [docs](https://cli.github.com/manual/gh_help_environment)).

However, it seems like the default `secrets.GITHUB_TOKEN` is not sufficient for our needs. It is limited to just the repository scope and cannot access projects or create/update issues etc.

@phlax would it be appropriate to pass the `${{ steps.appauth.outputs.token }}` token from the upstream `envoy-sync.yml` workflow, as an input to the downstream `envoy-sync-receive.yaml` workflow?

[the remark this answers is not in the document]
Comment (phlax) [reply ci/envoy-sync-receive.sh:77]: >  It is limited to just the repository scope and cannot access projects or create/update issues etc

i think you need

```yaml
permissions:
  issues: write
  pull-requests: write
```

creating/updating issues/prs should work with that 

re appauth - i think we dont need it at all if we are using the `GITHUB_TOKEN`

[the remark this answers is not in the document]
Comment (phlax) [reply ci/envoy-sync-receive.sh:77]: (that is stil repo-scoped tho)

[the remark this answers is not in the document]
Comment (tedjpoole) [reply ci/envoy-sync-receive.sh:77]: Thank you! Adding the `issues: write` permission enabled the `gh issue create` and `gh issue comment` commands to work.

However, despite adding the `pull-requests: write` permission, the `gh pr create` command still fails to create a pull request:
```
++ gh pr create --head auto-merge-main --base main --title 'auto-merge envoyproxy/envoy[main] into tedjpoole/test-downstream[main]' --body 'Generated by envoy-sync-receive.sh'
pull request create failed: GraphQL: GitHub Actions is not permitted to create or approve pull requests (createPullRequest)
```

Also, while testing locally I encountered this error from the `git push` command. It seems to be caused by the fact that the merge had brought in some upstream changes to workflow files under the `.github/workflows/` directory, and when pushing any changes to workflow files, github requires the actor to have `workflows: write` permission, which cannot be given to the `secrets.GITHUB_TOKEN`.
```
+ git push --force origin HEAD:auto-merge-main
To https://github.com/tedjpoole/test-downstream
 ! [remote rejected]       HEAD -> auto-merge-main (refusing to allow a GitHub App to create or update workflow `.github/workflows/_cache.yml` without `workflows` permission)
error: failed to push some refs to 'https://github.com/tedjpoole/test-downstream'
```

Therefore, it does seem that the `secrets.GITHUB_TOKEN` _cannot_ be given some of the permissions that we need in the script.
I know I need to use a different token which has the additional permissions, but I'm not really sure of the best way to achieve that.

If I reinstate this step:
```
    - id: appauth
      uses: envoyproxy/toolshed/gh-actions/appauth@actions-v0
      with:
        key: ${{ secrets.ENVOY_CI_UPDATE_BOT_KEY }}
        app_id: ${{ secrets.ENVOY_CI_UPDATE_APP_ID }}
```
then I think I can use `${{ steps.appauth.outputs.value }}` in place of `secrets.GITHUB_TOKEN`, but I'm not sure where I need to create those two secrets, or even if a am able to, nor where I can configure the permissions that are required by the script.

Any advice or pointers welcome thanks

File: .github/workflows/envoy-sync-receive.yaml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    - id: appauth
+      uses: envoyproxy/toolshed/gh-actions/appauth@actions-v0
+      with:
+        key: ${{ secrets.ENVOY_CI_UPDATE_BOT_KEY }}
+        app_id: ${{ secrets.ENVOY_CI_UPDATE_APP_ID }}
+    - name: "Checkout ${{ github.repository }}[${{ github.event.inputs.branch }}]"
+      uses: actions/checkout@a5ac7e51b41094c92402da3b24376905380afc29  # v4.1.6
+      with:
+        ref: ${{ github.event.inputs.branch }}
+        token: ${{ steps.appauth.outputs.value }}
```
Comment (phlax) [inline .github/workflows/envoy-sync-receive.yaml:36]:
```suggestion
        token: ${{ steps.appauth.outputs.token }}
```

File: .github/workflows/envoy-sync-receive.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        fetch-depth: 0
+    - name: "Set git user details"
+      run: |
+        git config user.name "${{ github.actor }}"
+        git config user.email "${{ github.actor_id }}+${{ github.actor }}@users.noreply.github.com"
+    - run: ci/envoy-sync-receive.sh ${{ github.event.inputs.branch }}
+      env:
+        GH_TOKEN: ${{ steps.appauth.outputs.value }}
```
Comment (phlax) [inline .github/workflows/envoy-sync-receive.yaml:44]:
```suggestion
        GH_TOKEN: ${{ steps.appauth.outputs.token }}
```

**Review by phlax — APPROVED** [review]
lgtm, thanks @tedjpoole 

nit re `appauth` - better to use `token` - see here:

https://github.com/envoyproxy/toolshed/blob/7f22091712156acf153d96e97dfd4ba9c11cb7d1/gh-actions/appauth/main.ts#L57-L59

**Pull request merged**